### PR TITLE
feat(data): add provideEntityData and provideEntityDataWithoutEffects

### LINKED ISSUE #3529: Data: Introduce Standalone API for NgRx Data
### Information

Introduce standalone APIs for using the `@ngrx/data` package. This should allow developers to use NgRx Data in a "module-free" application.

Based on the discussion in #3399

```ts
provideEntityData(config)
```

This function will make use of the [ENVIRONMENT_INITIALIZER](https://angular.io/api/core/ENVIRONMENT_INITIALIZER) token to perform any init logic that was previously done in the NgModule constructor.

### Describe any alternatives/workarounds you're currently using

Using `importProvidersFrom` function:

```ts
// main.ts
bootstrapApplication(AppComponent, {
  providers: [importProvidersFrom(EntityDataModule.forRoot(entityConfig))]
});
```

### I would be willing to submit a PR to fix this issue

- [ ] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by santoshyadavdev**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngrx/platform/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

Closes #3529 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: modules/data/src/entity-data.module.ts [hunk 1]
```diff
@@ -2,60 +2,25 @@ import { ModuleWithProviders, NgModule } from '@angular/core';
...
-    EntityCacheEffects,
-    EntityEffects,
-    { provide: HttpUrlGenerator, useClass: DefaultHttpUrlGenerator },
-    {
-      provide: PersistenceResultHandler,
-      useClass: DefaultPersistenceResultHandler,
-    },
-    { provide: Pluralizer, useClass: DefaultPluralizer },
-  ],
+  providers: [ENTITY_DATA_WITHOUT_EFFECTS_PROVIDERS, ENTITY_DATA_PROVIDERS],
```
Comment (markostanimirovic) [inline modules/data/src/entity-data.module.ts:23]:
```suggestion
  imports: [EntityDataModuleWithoutEffects],
  providers: [ENTITY_DATA_PROVIDERS],
```

btw, there are unused imports in this file (EffectsModule, EntityEffects, etc.) that should be removed

**Review by markostanimirovic — COMMENTED** [review]
Thanks a lot Santosh for reopening this PR!

**Comment by markostanimirovic** [comment]
I didn't add TSDoc and tests for the new data APIs yesterday, but it would be good to cover at least the main use cases. If you don't have time, we can add tests in a separate PR.

I'd also revert changes from `standalone-app`. We could improve the `data-example-app` project in the future and use standalone APIs there.

**Comment by santoshyadavdev** [comment]
> a

**Comment by markostanimirovic** [comment]
Hey @santoshyadavdev,

We don't need to exclude existing `@ngrx/data` tests. [The suggestion above](https://github.com/ngrx/platform/pull/3647#discussion_r1013452452) should fix existing tests (`imports: [EntityDataModuleWithoutEffects]` is currently missing).

**Comment by santoshyadavdev** [comment]
> Hey @santoshyadavdev,
> 
> We don't need to exclude existing `@ngrx/data` tests. [The suggestion above](https://github.com/ngrx/platform/pull/3647#discussion_r1013452452) should fix existing tests (`imports: [EntityDataModuleWithoutEffects]` is currently missing).

oops, my bad, really sorry, I am too tired 😂

File: modules/data/spec/entity-services/entity-collection-service.spec.ts [hunk 2]
```diff
@@ -29,6 +29,7 @@ import {
   Logger,
 } from '../..';
 
+// TODO fix these tests
```
Comment (markostanimirovic) [inline modules/data/spec/entity-services/entity-collection-service.spec.ts:32]:
```suggestion
```

File: modules/data/spec/entity-services/entity-collection-service.spec.ts [hunk 3]
```diff
@@ -133,9 +134,16 @@ describe('EntityCollectionService', () => {
...
+      heroCollectionService.getWithQuery({ name: 'foo' }).subscribe(
+        () => {
+          console.log('expected error', error);
+          expectErrorToBe(error, done);
+        },
+        (err) => {
+          console.log('actual error', err);
+          expectErrorToBe(error, done);
+        }
+      );
```
Comment (markostanimirovic) [inline modules/data/spec/entity-services/entity-collection-service.spec.ts:146]: Only this test is currently failing. To make it work, we need to revert this change

File: projects/standalone-app-e2e/src/support/app.po.ts [hunk 4]
```diff
@@ -1,2 +1,2 @@
 export const getGreeting = () => cy.get('h1');
-export const loadFeature = () => cy.get('a').click();
+export const loadFeature = () => cy.get('a').contains('Load Feature').click();
```
Comment (markostanimirovic) [inline projects/standalone-app-e2e/src/support/app.po.ts:2]: This change should be also reverted

**Review by markostanimirovic — COMMENTED** [review]
> oops, my bad, really sorry, I am too tired 😂

Oh 😅 No worries!

**Review by markostanimirovic — APPROVED** [review]
🚀

**Comment by brandonroberts** [comment]
Changes look good overall. We have an opportunity to change the `provideEntityDataWithoutEffects()` to a better name.

I propose two options

1. Rename to `provideBaseEntityData()` and keep `provideEntityData()`

2. Use the pattern similar to the Angular Router

```ts
provideEntityData(config, [withEffects()])
```

Option 2 would be the default

**Comment by markostanimirovic** [comment]
> Changes look good overall. We have an opportunity to change the `provideEntityDataWithoutEffects()` to a better name.
> 
> I propose two options
> 
> 1. Rename to `provideBaseEntityData()` and keep `provideEntityData()`
> 2. Use the pattern similar to the Angular Router
> 
> ```ts
> provideEntityData(config, [withEffects()])
> ```
> 
> Option 2 would be the default

Sounds good to me 👍 I'm also not a big fan of `...withoutEffects` name 😅

**Review by brandonroberts — APPROVED** [review]
LGTM.

We can change the signature in a follow-up PR

**Comment by santoshyadavdev** [comment]
I can make changes tonight, if you want to wait 😃

**Comment by brandonroberts** [comment]
Thanks Santosh! I think it's ok to go ahead and land this one. @markostanimirovic is going to look at the follow-up PR

**Pull request merged**